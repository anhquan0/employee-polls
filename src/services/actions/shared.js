import {receiveUsers} from "./users";
import {receive} from "./questions";
import {getInitialData} from "../../utils/api";

export function initialData() {
    return async (dispatch) => {
        const { users, questions } = await getInitialData();
        dispatch(receiveUsers(users));
        dispatch(receive(questions));
    };
}
