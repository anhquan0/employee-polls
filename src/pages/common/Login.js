import React, { useState } from 'react';
import { connect } from "react-redux";
import { Navigate } from "react-router-dom";
import { handleLogin } from "../../services/actions/authedUser";
import PropTypes from "prop-types";
import '../../css/login.css';

const users = [
  { id: "quanlna1", name: "Quan Le Ngoc Anh", password: "1234567" },
  { id: "hiennt260", name: "Hien Nguyen Thuy", password: "1234567" },
  { id: "johndoe", name: "John Doe", password: "123123" },
  { id: "zoshikanlu", name: "Zenobia Oshikanlu", password: "pass246" },
];

const Login = ({ dispatch, loggedIn }) => {
  const [selectedUser, setSelectedUser] = useState(users[0]);

  if (loggedIn) {
    const urlParams = new URLSearchParams(window.location.search);
    const redirectUrl = urlParams.get("redirectTo");
    return <Navigate to={redirectUrl ? redirectUrl : "/"} />;
  }

  const handleUserChange = (e) => {
    const user = users.find((u) => u.id === e.target.value);
    setSelectedUser(user);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!selectedUser.id || !selectedUser.password) {
      alert("Please select a user!");
      return;
    }
    dispatch(handleLogin(selectedUser.id, selectedUser.password));
  };

  return (
    <div className="login-container">
      <div className="login-form">
        <h2 className="login-heading" data-testid="login-heading">Sign In</h2>
        <form onSubmit={handleSubmit}>
          <select
            className="user-select"
            value={selectedUser.id}
            onChange={handleUserChange}
            data-testid="user-select"
          >
            {users.map((user) => (
              <option key={user.id} value={user.id}>
                {user.name}
              </option>
            ))}
          </select>
          <button type="submit" className="login-button" data-testid="loginBtn">
            Log In
          </button>
        </form>
      </div>
    </div>
  );
};

Login.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loggedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = ({ authedUser }) => ({
  loggedIn: !!authedUser,
});

export default connect(mapStateToProps)(Login);