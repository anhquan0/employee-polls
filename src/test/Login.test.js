import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../services/store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Login from "../pages/common/Login";
import { initialData } from "../services/actions/shared";

describe("Login", () => {
    it("Renders Login component successfully", () => {
        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it('Handles login form submission correctly', async () => {
        await store.dispatch(initialData());

        const wrapper = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login/>
                </BrowserRouter>
            </Provider>
        );

        const loginHeadingElement = wrapper.getByTestId("login-heading");
        const submitButtonElement = wrapper.getByTestId("loginBtn");
        const userSelectElement = wrapper.getByTestId("user-select");

        expect(loginHeadingElement).toBeInTheDocument();
        expect(submitButtonElement).toBeInTheDocument();
        expect(userSelectElement).toBeInTheDocument();

        // Test user selection
        fireEvent.change(userSelectElement, { target: { value: 'johndoe' } });
        expect(userSelectElement.value).toBe("johndoe");

        // Test form submission
        fireEvent.click(submitButtonElement);
        
        // The component should redirect after successful login
        // We don't need to test for clearing fields since we're using a select
    });
});