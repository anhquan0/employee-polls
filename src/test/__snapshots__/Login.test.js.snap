// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Render sucess Login component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <div
          class="login-form"
        >
          <h2
            class="login-heading"
            data-testid="login-heading"
          >
            Sign In
          </h2>
          <form>
            <select
              class="user-select"
              data-testid="user-select"
            >
              <option
                value="quanlna1"
              >
                Quan Le Ngoc Anh
              </option>
              <option
                value="hiennt260"
              >
                Hien Nguyen Thuy
              </option>
              <option
                value="johndoe"
              >
                John Doe
              </option>
              <option
                value="zoshikanlu"
              >
                Zenobia Oshikanlu
              </option>
            </select>
            <button
              class="login-button"
              data-testid="loginBtn"
              type="submit"
            >
              Log In
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <div
        class="login-form"
      >
        <h2
          class="login-heading"
          data-testid="login-heading"
        >
          Sign In
        </h2>
        <form>
          <select
            class="user-select"
            data-testid="user-select"
          >
            <option
              value="quanlna1"
            >
              Quan Le Ngoc Anh
            </option>
            <option
              value="hiennt260"
            >
              Hien Nguyen Thuy
            </option>
            <option
              value="johndoe"
            >
              John Doe
            </option>
            <option
              value="zoshikanlu"
            >
              Zenobia Oshikanlu
            </option>
          </select>
          <button
            class="login-button"
            data-testid="loginBtn"
            type="submit"
          >
            Log In
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login Renders Login component successfully 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <div
          class="login-form"
        >
          <h2
            class="login-heading"
            data-testid="login-heading"
          >
            Sign In
          </h2>
          <form>
            <select
              class="user-select"
              data-testid="user-select"
            >
              <option
                value="quanlna1"
              >
                Quan Le Ngoc Anh
              </option>
              <option
                value="hiennt260"
              >
                Hien Nguyen Thuy
              </option>
              <option
                value="johndoe"
              >
                John Doe
              </option>
              <option
                value="zoshikanlu"
              >
                Zenobia Oshikanlu
              </option>
            </select>
            <button
              class="login-button"
              data-testid="loginBtn"
              type="submit"
            >
              Log In
            </button>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <div
        class="login-form"
      >
        <h2
          class="login-heading"
          data-testid="login-heading"
        >
          Sign In
        </h2>
        <form>
          <select
            class="user-select"
            data-testid="user-select"
          >
            <option
              value="quanlna1"
            >
              Quan Le Ngoc Anh
            </option>
            <option
              value="hiennt260"
            >
              Hien Nguyen Thuy
            </option>
            <option
              value="johndoe"
            >
              John Doe
            </option>
            <option
              value="zoshikanlu"
            >
              Zenobia Oshikanlu
            </option>
          </select>
          <button
            class="login-button"
            data-testid="loginBtn"
            type="submit"
          >
            Log In
          </button>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
